File: go/day19_part1_2021.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	file, err := os.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}
	input := strings.TrimSpace(string(file))

	result := solve(input)
	fmt.Println(result)
}

func solve(input string) int {
	scanners := parseInput(input)

	settled := []scanner{scanners[0]}
	settled[0].absoluteCoords = settled[0].relativeCoords
	settled[0].fillAbsoluteCoordsMap()

	undetermined := scanners[1:]

	for len(undetermined) > 0 {
		for i, undet := range undetermined {
			maybeUpdated, ok := findAbsoluteCoordsForScanner(undet, settled)
			if ok {
				settled = append(settled, maybeUpdated)

				copy(undetermined[i:], undetermined[i+1:])

				undetermined = undetermined[:len(undetermined)-1]

				break
			}
		}
	}

	allBeacons := map[[3]int]bool{}
	for _, s := range settled {
		for c := range s.absoluteCoordsMap {
			allBeacons[c] = true
		}
	}

	return len(allBeacons)
}

type scanner struct {
	number            int
	x, y, z           int
	relativeCoords    [][3]int
	rotations         [][][3]int
	absoluteCoords    [][3]int
	absoluteCoordsMap map[[3]int]bool
}

func (s *scanner) fillAbsoluteCoordsMap() {
	s.absoluteCoordsMap = map[[3]int]bool{}
	if len(s.absoluteCoords) == 0 {
		panic(fmt.Sprintf("absolute coords not set for scanner %d", s.number))
	}
	for _, ac := range s.absoluteCoords {
		s.absoluteCoordsMap[ac] = true
	}
}

func (s *scanner) fillRotations() {

	posX := s.relativeCoords
	var dir2, dir3, dir4, dir5, dir6 [][3]int
	for _, c := range posX {
		x, y, z := c[0], c[1], c[2]
		dir2 = append(dir2, [3]int{x, -y, -z})
		dir3 = append(dir3, [3]int{x, -z, y})
		dir4 = append(dir4, [3]int{-y, -z, x})
		dir5 = append(dir5, [3]int{-x, -z, -y})
		dir6 = append(dir6, [3]int{y, -z, -x})
	}
	sixRotations := [][][3]int{
		posX, dir2,
		dir3, dir4,
		dir5, dir6,
	}

	var finalRotations [][][3]int
	for _, rotation := range sixRotations {
		var r2, r3, r4 [][3]int
		for _, c := range rotation {
			x, y, z := c[0], c[1], c[2]
			r2 = append(r2, [3]int{-y, x, z})
			r3 = append(r3, [3]int{-x, -y, z})
			r4 = append(r4, [3]int{y, -x, z})
		}
		finalRotations = append(finalRotations, rotation, r2, r3, r4)
	}
	s.rotations = finalRotations
}

func findAbsoluteCoordsForScanner(undet scanner, settled []scanner) (maybeUpdated scanner, didUpdate bool) {

	for _, rotatedCoords := range undet.rotations {

		for _, set := range settled {
			for _, absCoord := range set.absoluteCoords {

				for _, relativeCoord := range rotatedCoords {

					unsettledAbsoluteCoords := makeAbsoluteCoordsList(absCoord, relativeCoord, rotatedCoords)

					var matchingCount int

					for _, ac := range unsettledAbsoluteCoords {
						if set.absoluteCoordsMap[ac] {

							matchingCount++
						}
					}

					if matchingCount >= 12 {
						undet.relativeCoords = rotatedCoords
						undet.absoluteCoords = unsettledAbsoluteCoords
						undet.fillAbsoluteCoordsMap()
						undet.x = absCoord[0] - relativeCoord[0]
						undet.y = absCoord[1] - relativeCoord[1]
						undet.z = absCoord[2] - relativeCoord[2]
						return undet, true
					}
				}
			}
		}
	}

	return undet, false
}

func makeAbsoluteCoordsList(absolute, relative [3]int, relativeCoords [][3]int) [][3]int {

	diff := [3]int{
		absolute[0] - relative[0],
		absolute[1] - relative[1],
		absolute[2] - relative[2],
	}

	var absCoords [][3]int
	for _, c := range relativeCoords {
		absCoords = append(absCoords, [3]int{
			diff[0] + c[0],
			diff[1] + c[1],
			diff[2] + c[2],
		})
	}

	return absCoords
}

func parseInput(input string) (ans []scanner) {
	for _, rawScanner := range strings.Split(input, "\n\n") {
		var number int
		lines := strings.Split(rawScanner, "\n")
		_, err := fmt.Sscanf(lines[0], "--- scanner %d ---", &number)
		if err != nil {
			panic("parsing error " + err.Error())
		}

		var coords [][3]int
		for _, line := range lines[1:] {
			var x, y, z int
			_, err := fmt.Sscanf(line, "%d,%d,%d", &x, &y, &z)
			if err != nil {
				panic("parsing error " + err.Error())
			}
			coords = append(coords, [3]int{x, y, z})
		}

		sc := scanner{
			number:            number,
			x:                 0,
			y:                 0,
			z:                 0,
			relativeCoords:    coords,
			absoluteCoords:    nil,
			absoluteCoordsMap: map[[3]int]bool{},
		}
		sc.fillRotations()
		ans = append(ans, sc)
	}

	return ans
}

File: julia/day19_part1_2021.jl

function solve()
    input = read("input.txt", String)
    scanners = parse_input(input)
    settled = [scanners[1]]
    settled[1].absolute_coords = settled[1].relative_coords
    fill_absolute_coords_map!(settled[1])
    undetermined = scanners[2:end]
    while !isempty(undetermined)
        for (i, undet) in enumerate(undetermined)
            maybe_updated, ok = find_absolute_coords_for_scanner(undet, settled)
            if ok
                push!(settled, maybe_updated)
                deleteat!(undetermined, i)
                break
            end
        end
    end
    all_beacons = Set{Tuple{Int, Int, Int}}()
    for s in settled
        for c in keys(s.absolute_coords_map)
            push!(all_beacons, c)
        end
    end
    return length(all_beacons)
end

mutable struct Scanner
    number::Int
    x::Int
    y::Int
    z::Int
    relative_coords::Vector{Tuple{Int, Int, Int}}
    rotations::Vector{Vector{Tuple{Int, Int, Int}}}
    absolute_coords::Vector{Tuple{Int, Int, Int}}
    absolute_coords_map::Dict{Tuple{Int, Int, Int}, Bool}
end

function fill_absolute_coords_map!(s::Scanner)
    s.absolute_coords_map = Dict{Tuple{Int, Int, Int}, Bool}()
    if isempty(s.absolute_coords)
        error("absolute coords not set for scanner $(s.number)")
    end
    for ac in s.absolute_coords
        s.absolute_coords_map[ac] = true
    end
end

function fill_rotations!(s::Scanner)
    pos_x = s.relative_coords
    dir2 = map(c -> (c[1], -c[2], -c[3]), pos_x)
    dir3 = map(c -> (c[1], -c[3], c[2]), pos_x)
    dir4 = map(c -> (-c[2], -c[3], c[1]), pos_x)
    dir5 = map(c -> (-c[1], -c[3], -c[2]), pos_x)
    dir6 = map(c -> (c[2], -c[3], -c[1]), pos_x)
    six_rotations = [pos_x, dir2, dir3, dir4, dir5, dir6]
    final_rotations = Vector{Vector{Tuple{Int, Int, Int}}}()
    for rotation in six_rotations
        r2 = map(c -> (-c[2], c[1], c[3]), rotation)
        r3 = map(c -> (-c[1], -c[2], c[3]), rotation)
        r4 = map(c -> (c[2], -c[1], c[3]), rotation)
        append!(final_rotations, [rotation, r2, r3, r4])
    end
    s.rotations = final_rotations
end

function find_absolute_coords_for_scanner(undet::Scanner, settled::Vector{Scanner})
    for rotated_coords in undet.rotations
        for set in settled
            for abs_coord in set.absolute_coords
                for relative_coord in rotated_coords
                    unsettled_absolute_coords = make_absolute_coords_list(abs_coord, relative_coord, rotated_coords)
                    matching_count = 0
                    for ac in unsettled_absolute_coords
                        if haskey(set.absolute_coords_map, ac)
                            matching_count += 1
                        end
                    end
                    if matching_count >= 12
                        undet.relative_coords = rotated_coords
                        undet.absolute_coords = unsettled_absolute_coords
                        fill_absolute_coords_map!(undet)
                        undet.x = abs_coord[1] - relative_coord[1]
                        undet.y = abs_coord[2] - relative_coord[2]
                        undet.z = abs_coord[3] - relative_coord[3]
                        return undet, true
                    end
                end
            end
        end
    end
    return undet, false
end

function make_absolute_coords_list(absolute::Tuple{Int, Int, Int}, relative::Tuple{Int, Int, Int}, relative_coords::Vector{Tuple{Int, Int, Int}})
    diff = (absolute[1] - relative[1], absolute[2] - relative[2], absolute[3] - relative[3])
    abs_coords = map(c -> (diff[1] + c[1], diff[2] + c[2], diff[3] + c[3]), relative_coords)
    return abs_coords
end

function parse_input(input::String)
    scanners = Scanner[]
    for raw_scanner in split(input, "\n\n")
        lines = split(raw_scanner, "\n")
        number = parse(Int, match(r"--- scanner (\d+) ---", lines[1])[1])
        coords = map(lines[2:end]) do line
            x, y, z = map(x -> parse(Int, x), split(line, ","))
            (x, y, z)
        end
        sc = Scanner(number, 0, 0, 0, coords, Vector{Vector{Tuple{Int, Int, Int}}}(), Vector{Tuple{Int, Int, Int}}(), Dict{Tuple{Int, Int, Int}, Bool}())
        fill_rotations!(sc)
        push!(scanners, sc)
    end
    return scanners
end

println(solve())
